File: Iteration2/src/Login/AdvisorLoginStrategy.java
package Login;

import Domain.Advisor;
import Domain.Department;
import Domain.RegistrationStatus;

import java.util.Scanner;

public class AdvisorLoginStrategy implements ILoginStrategy {
    @Override
    public void login(Department department) {
        loginAsAnAdvisor(department);
    }

    private void loginAsAnAdvisor(Department department) {
        short var = -1;
        boolean flag = false;
        Advisor advisor = null;
        LoginHelper loginHelper = new LoginHelper();

        while (var != 0 && !flag) {
            System.out.println("Enter your staff number");
            System.out.println("0 to cancel");

            String enteredNo;


            try{
                Scanner input2 = new Scanner(System.in);
                enteredNo = input2.next();
            }catch (Exception e){
                System.out.println(e);
                break;
            }
            if(enteredNo.equals("0")){
                return;
            }
            advisor = department.getAdvisorByStaffNo(enteredNo);
            if (advisor == null) {
                continue;
            }
            flag = true;
        }

        short choice = -1;
        while (choice != 0) {
            System.out.println("0 to exit");
            System.out.println("1 to list active registrations");
            System.out.println("2 to list students you advise");
            System.out.println("3 to proceed a registration");
            System.out.println("4 to list teaching courses");
            System.out.println("5 to grade ");
            System.out.println("");
            try{
                Scanner input2 = new Scanner(System.in);
                choice = input2.nextShort();
            }catch (Exception e){
                System.out.println(e);
            }

            switch (choice) {
                case 0:
                    break;
                case 1:                                 //get active registrations
                    System.out.println(advisor.getActiveRegistrations());
                    break;
                case 2:                                 //list students you advise
                    for (String code : advisor.getStudentNumbers()) {
                        System.out.println(department.getStudentByStudentNo(code));
                    }
                    break;
                case 3:                                     //proceed a registration
                    String registrationNo, registrationStatusStr;
                    try{
                        Scanner input = new Scanner(System.in);
                        System.out.println("Enter registration no.");
                        registrationNo = input.next();
                        System.out.println("Enter status(Active-Rejected-Confirmed)");
                        registrationStatusStr = input.next();
                    }catch (Exception e){
                        System.out.println(e);
                        break;
                    }
                    RegistrationStatus status = null;
                    if (registrationStatusStr.equalsIgnoreCase("Confirmed")) {
                        status = RegistrationStatus.Confirmed;
                    } else if (registrationStatusStr.equalsIgnoreCase("Active")) {
                        status = RegistrationStatus.Active;
                    } else if (registrationStatusStr.equalsIgnoreCase("Rejected")) {
                        status = RegistrationStatus.Rejected;
                    } else {
                        System.out.println("Invalid registration status!");
                        break;
                    }
                    advisor.proceedTheRegistration(registrationNo, status);
                    break;
                case 4:
                    loginHelper.listCoursesForLecturer(department, advisor);
                default://default case
                    break;
            }

        }//While loop end
    }//Login as an advisor end
}

File: Iteration3/Login/AdvisorLoginStrategy.py
from Domain.RegistrationStatus import RegistrationStatus
from Login.ILoginStrategy import ILoginStrategy
from Login.LoginHelper import LoginHelper
import logging

class AdvisorLoginStrategy(ILoginStrategy):
    def login(self, department):
        self.login_as_an_advisor(department)

    def login_as_an_advisor(self, department):
        logger = logging.getLogger()
        var = -1
        flag = False
        advisor = None
        login_helper = LoginHelper()

        while var != 0 and not flag:
            print("Enter your staff number")
            print("0 to cancel")

            entered_no = input()

            if entered_no == "0":
                return

            advisor = department.get_advisor_by_staff_no(entered_no)
            if advisor is None:
                logger.info(f"Advisor login attempt unsuccesfull. Advisor not found with {entered_no}")

                continue
            else:
                print("Enter your password:")
                entered_password = input()
                if entered_password != advisor.get_password():
                    logger.info(f"Advisor login attempt unsuccesfull. Advisor with {entered_no} entered wrong password")
                    print("Wrong password try again")
                    continue
            flag = True
        logger.info(f"Advisor login attempt successfull. Advisor with {entered_no} loginned")

        choice = -1
        while choice != 0:
            print("0 to exit")
            print("1 to list active registrations")
            print("2 to list students you advise")
            print("3 to proceed a registration")
            print("4 to list teaching courses")
            print("5 to grade ")
            print("")
            try:
                choice = int(input())
            except ValueError as e:
                print(e)

            if choice == 0:
                break
            elif choice == 1:  # get active registrations
                for registration in advisor.get_active_registrations():
                    print(registration)
            elif choice == 2:  # list students you advise
                for code in advisor.get_student_numbers():
                    print(department.get_student_by_student_no(code))
            elif choice == 3:  # proceed a registration
                try:
                    registration_no = input("Enter registration no.: ")
                    registration_status_str = input("Enter status (Active-Rejected-Confirmed): ")
                except Exception as e:
                    print(e)
                    logger.info(f"{e} for advisor {entered_no}")
                    break

                status = None
                if registration_status_str.lower() == "confirmed":
                    status = RegistrationStatus.Confirmed
                    logger.info(f"Advisor {entered_no} confirmed registration: {registration}")
                elif registration_status_str.lower() == "active":
                    status = RegistrationStatus.Active
                    logger.info(f"Advisor {entered_no} did not change the status of registration: {registration}")
                elif registration_status_str.lower() == "rejected":
                    logger.info(f"Advisor {entered_no} rejected registration: {registration}")

                    status = RegistrationStatus.Rejected
                else:
                    print("Invalid registration status!")
                    logger.info(f"Advisor {entered_no} entered invalid registration status")
                    continue

                advisor.proceed_the_registration(registration_no, status)
            elif choice == 4:
                login_helper.list_courses_for_lecturer(department, advisor)
            else:
                # default case
                break
